### PR TITLE
Add an SLA for gcp-ingestion

### PR DESCRIPTION
**Pull request by relud**
Phrasing is largely modeled after https://cloud.google.com/pubsub/sla

Thresholds were made up based on what I think is reasonable to achieve by the end of Q4.

Requesting review from data-platform before soliciting feedback from operations.

### CODE REVIEW

**Comment by relud** [comment]
cc @sunahsuh

File: docs/sla.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+exceeds 1 hour for a batch sink job or 1 minute for a decoder or streaming sink
+job.
+
+**"Downtime Period"** means a period of 60 consecutive seconds of Downtime.
+Intermittent Downtime for a period of less than 60 consecutive seconds will not
+be counted towards any Downtime Periods.
+
+**"Biweekly Uptime Percentage"** means total number of minutes in two weeks,
+minus the number of minutes of Downtime suffered from all Downtime Periods in
+two weeks, divided by the total number of minutes in two weeks.
```
Comment (jklukas) [inline docs/sla.md]: Is this a rolling two weeks, or are we going to have some defined start date and end date for each biweekly period?

File: docs/sla.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+**"Stability Work"** means work on preventing downtime while all other features
+and work are suspended.
+
+**"Stability Work Period"** means a period of time spent on Stability Work
+determined below.
+
+| Covered Service | Biweekly Uptime Percentage | Time spent on Stability Work |
```
Comment (jklukas) [inline docs/sla.md]: Is the "time spent" here referring to a team's time? Or to person-weeks of effort? So a promise of "3 months" of time spent could potentially be 3 engineers dedicated for 1 month?

File: docs/sla.md (hunk 1 above)
Comment (relud) [reply docs/sla.md]: rolling period. i'll update this to say that.

File: docs/sla.md (hunk 2 above)
Comment (relud) [reply docs/sla.md]: it's "time spent" as in a dedicated "3 months" window that applies to relevant people. so if it's three engineers, they're still dedicated for 3 months.

Comment (relud) [reply docs/sla.md]: also, it may be 3 months for one person, and another person assigned for review.

Comment (relud) [reply docs/sla.md]: Also, this is not intended to indicate that time can't be taken to task switch as a fatigue reducing measure. Just that this is the only assigned work for the period, and that it is the only work that should be deployed to production during the period.

**Comment by relud** [comment]
> if we're going to merge this to the repo in the near term, I think there needs to be a disclaimer at the top that the contract here is currently aspirational.

this won't be merged unless operations and management agree to the terms, which I very much doubt will happen anywhere near quickly or without modification.

I think it would be very good to have a highlighted disclaimer indicating that this is a prototype SLA as an agreement between the developers/managers and themselves, and is not an agreement, implied or otherwise, with anyone outside that group. I think it would also be good to include something like a mission statement for the SLA, which in this case is primarily to respond to downtime with behavior that reduces future downtime and secondarily to establish internal expectations regarding downtime. I'll try to write something like that up and add it.

> We need to understand how we'll be measuring all the quantities here

for the server error and message age, I'm confident it will be available via stackdriver. I think that it's something operations will be in charge of measuring, and I chose the language there specifically based on what I thought would be reasonable to evaluate automatically.

I have no idea how we should represent the downtime due to misbehaving code. I chose effecting `at least 0.1%` of traffic, but now that I think about it what I was really trying to express was a discrepancy requiring backfill (as opposed to requiring modification of the spec). Downtime would be measured relative to the period that requires backfill, but that's actually a proxy for measuring how much the issue impacted or could have impacted external parties.

I think that misbehaving code will cause far longer downtime periods. I'm uncertain about how handle how that relates to the "Stability Work Period", because I don't think every beam job bug that goes unnoticed for more than 2 days should realistically result in dropping non-stability quarterly goals. On the other hand I feel like that would be very effective at minimizing back fills due to code here.

> who is accountable for confirming and enforcing the SLA

Stackdriver can monitor the non-code portion of the SLA and send out an email indicating a breach of SLA and the expected response. I expect we should be able to set up an integration between stackdriver and statuspage to make it more visible.

We would have a postmortem as normal, so for enforcement I expect that would include checking in to determine which engineers it would be most effective and appropriate to assign a "Stability Work Period" to minimize future downtime. I also expect this would implicitly include re-evaluating the documented length of the "Stability Work Period" if it is deemed inappropriately long or short.

> we should be extra careful about what we're messaging at this stage

Agreed, thorough and careful communication about this is vital to avoiding unintentional side-effects.

**Comment by relud** [comment]
I said: 

> I have no idea how we should represent the downtime due to misbehaving code. I chose effecting `at least 0.1%` of traffic, but now that I think about it what I was really trying to express was a discrepancy requiring backfill (as opposed to requiring modification of the spec). Downtime would be measured relative to the period that requires backfill, but that's actually a proxy for measuring how much the issue impacted or could have impacted external parties.

I didn't change it though, because I'm having trouble finding wording that is concise and unambiguous.

**Comment by relud** [comment]
@whd I'd like to invite you to review at this point. I'm specifically interested in your personal opinion on the document.

Afterward (still before merging) I'll schedule a meeting with relevant parties to discuss things like manager agreement and how this effects monitoring and alerting.

File: docs/sla.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+| Covered Service | Biweekly Uptime Percentage |
+|-----------------|----------------------------|
+| ingestion-edge  | 99.99%                     |
+| ingestion-beam  | 99.5%                      |
+
+## Definitions
+
+**"Covered Service"** means ingestion-edge and ingestion-beam jobs.
+
+**"Downtime"** on ingestion-edge means for at least 0.1% of requests [the
```
Comment (whd) [inline docs/sla.md]: The spec was not "followed" is vague as to what is doing the following, and is potentially vague as to the meaning of "following".

**Comment by relud** [comment]
this has some external issues, i'm going to close this pr for now, and keep the branch around for later use if needed.

**Pull request closed without merging**